src/borg/helpers/time.py

```diff
@@ -185,3 +193,298 @@ def isoformat(self):
+
+
+class DatePatternError(ValueError):
+    """Raised when a date: archive pattern cannot be parsed."""
+
+
+def exact_predicate(dt: datetime):
+    """Return predicate matching archives whose ts equals dt (UTC)."""
+    dt_utc = dt.astimezone(timezone.utc)
+    return lambda ts: ts.astimezone(timezone.utc) == dt_utc
+
+
+def interval_predicate(start: datetime, end: datetime):
+    start_utc = start.astimezone(timezone.utc)
+    end_utc = end.astimezone(timezone.utc)
+    if start_utc > end_utc:
+        raise DatePatternError("start date must be before end date")
+    return lambda ts: start_utc <= ts.astimezone(timezone.utc) < end_utc
+
+
+def parse_tz(tzstr: str):
+    """
+    Parses a UTC offset like +08:00 or [Region/Name] into a timezone object.
+    """
+    if not tzstr:
+        return None
+    if tzstr == "Z":
+        return timezone.utc
+    if tzstr[0] in "+-":
+        sign = 1 if tzstr[0] == "+" else -1
+        try:
+            hh, mm = map(int, tzstr[1:].split(":"))
+            if not (0 <= mm < 60):
+                raise ValueError
+        except Exception:
+            raise DatePatternError("invalid UTC offset format")
+        # we do it this way so that, for example, -8:30 is
+        # -8 hours and -30 minutes, not -8 hours and +30 minutes
+        total_minutes = sign * (hh * 60 + mm)
+        # enforce ISO-8601 bounds (-12:00 to +14:00)
+        if not (-12 * 60 <= total_minutes <= 14 * 60):
+            raise DatePatternError("UTC offset outside ISO-8601 bounds")
+        return timezone(timedelta(minutes=total_minutes))
+    # [Region/Name]
+    try:
+        return ZoneInfo(tzstr.strip("[]"))
+    except Exception:
+        raise DatePatternError("invalid timezone format")
+
+
+def build_datetime_from_groups(gd: dict, tz: timezone) -> datetime:
+    """
+    Construct a datetime from partial ISO groups, filling missing fields with
+    the earliest valid value, and attaching tzinfo.
+    """
+    year = int(gd["year"])
+    month = int(gd.get("month") or 1)
+    day = int(gd.get("day") or 1)
+    hour = int(gd.get("hour") or 0)
+    minute = int(gd.get("minute") or 0)
+    # handle fractional seconds
+    microsecond = 0
+    second = 0
+    sec_str = gd.get("second")
+    if sec_str:
+        if "." in sec_str:
+            whole, frac = sec_str.split(".", 1)
+            second = int(whole)
+            # pad or trim frac to microseconds
+            microsecond = int(float(f"0.{frac}") * 1_000_000)
+        else:
+            second = int(sec_str)
+    return datetime(year, month, day, hour, minute, second, microsecond, tzinfo=tz)
+
+
+# Regex for ISO-8601 timestamps:
+# Accepts both 'T' and space as separators between date and time per RFC-3339/IXDTF.
+MAIN_RE = r"""
+  ^
+  (?:
+     # ISO week date: YYYY-Www or YYYY-Www-D
+     (?P<isoweek_year>\d{4})-W(?P<isoweek_week>\d{2})(?:-(?P<isoweek_day>\d))?
+   | # Ordinal date: YYYY-DDD
+     (?P<ordinal_year>\d{4})-(?P<ordinal_day>\d{3})
```

Comment:
This might be a bit silly, but it just got me thinking: Could we support ordinal days with wildcard years? :thinking: Same for ISO weeks, possibly even `*-W*-5` (wow, that looks crazy :laughing:) to match all archives created on a Friday? Not sure whether users would use it, but if it's possible? WDYT?